uint256.go

```diff
@@ -198,9 +198,29 @@ func (z *Int) AddMod(x, y, m *Int) *Int {
 	if m.IsZero() {
 		return z.Clear()
 	}
-	if z == m { // z is an alias for m  // TODO: Understand why needed and add tests for all "division" methods.
+	if z == m { // z is an alias for m and will be overwritten by AddOverflow before m is read
 		m = m.Clone()
 	}
+
+	// Fast path for m >= 2^192, with x and y at most slightly bigger than m.
+	// This is always the case when x and y are already reduced modulo such m.
+
+	if (m[3] != 0) && (m[3] >= x[3]) && (m[3] >= y[3]) {
+		if z, overflow := z.AddOverflow(x, y); overflow {
+			z.Sub(z, m)
+		}
+
+		for {
+			t := *z
+			if _, overflow := t.SubOverflow(&t, m); overflow {
+				break
```

Comment:
How many times this loop can run?

Reply:
The worst case here is m = 2^192, x = y = 2m-1, so x+y = 4m-2. Hence the loop is exited at the latest in the fourth iteration. For already-reduced inputs, this happens no later than the second iteration.

Reply:
This comment should be added to the code.

Does this procedure has any name?

Reply:
Comment added.
This procedure doesn't have any particular name, it's just repeated subtraction of the modulus until we have the smallest non-negative residue.